### PR TITLE
Makefile: do not generate files at every build

### PR DESCRIPTION
**Pull request by lebauce**

### CODE REVIEW

**Comment by lebauce** [comment]
run skydive-unit-tests

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-elasticsearch

File: Makefile [hunk 1]
```diff
@@ -211,21 +205,36 @@ debug.analyzer:
 	sed -e 's/type Flow struct {/type Flow struct { XXX_state flowState `json:"-"`/' -i.bak flow/flow.pb.go
 	gofmt -s -w flow/flow.pb.go
 
-.PHONY: .easyjson.all
-.easyjson.all: builddep .proto ${EASYJSON_FILES_ALL}
-	$(call VENDOR_RUN,${EASYJSON_GITHUB}) easyjson -all ${EASYJSON_FILES_ALL}
+.proto: govendor flow/flow.pb.go filters/filters.pb.go http/wsstructmessage.pb.go
```
Comment (hunchback) [inline Makefile]:
```
PROTO_FILES = \
        flow/flow.pb.go \
        filters/filters.pb.go \
        http/wsstructmessage.pb.go

.PHONY: .proto
.proto: $(PROTO_FILES)

.PHONY: .proto.clean
.proto.clean:
        rm -rf $PROTO_FILES)
```

File: Makefile [hunk 2]
```diff
@@ -211,21 +205,36 @@ debug.analyzer:
...
+
+.PHONY: .proto.clean
+.proto.clean:
+	find . \( -name *.pb.go ! -path './vendor/*' \) -exec rm {} \;
+
+GEN_EASYJSON_FILES_TAG := $(patsubst %.go,%_easyjson.go,$(EASYJSON_FILES_TAG))
+GEN_EASYJSON_FILES_TAG_LINUX := $(patsubst %.go,%_easyjson.go,$(EASYJSON_FILES_TAG_LINUX))
+GEN_EASYJSON_FILES_TAG_OPENCONTRAIL := $(patsubst %.go,%_easyjson.go,$(EASYJSON_FILES_TAG_OPENCONTRAIL))
```
Comment (hunchback) [inline Makefile]:
```
define easyjson_tag
$(patsubst %.go,%_easyjson.go,$1)
endef

GEN_EASYJSON_FILES_TAG := $(call easyjson_tag,$(EASYSON_FILES_TAG))
...
```

File: Makefile [hunk 3]
```diff
@@ -211,21 +205,36 @@ debug.analyzer:
...
-.PHONY: .easyjson.tag.opencontrail
-.easyjson.tag.opencontrail: builddep ${EASYJSON_FILES_TAG_OPENCONTRAIL}
-	$(call VENDOR_RUN,${EASYJSON_GITHUB}) easyjson -build_tags "opencontrail" ${EASYJSON_FILES_TAG_OPENCONTRAIL}
+topology/probes/socketinfo/connection_easyjson.go: topology/probes/socketinfo/connection.go
+	$(call VENDOR_RUN,${EASYJSON_GITHUB}) easyjson -build_tags linux $<
+
+topology/probes/opencontrail/routing_table_easyjson.go: $(EASYJSON_FILES_TAG_OPENCONTRAIL)
+	$(call VENDOR_RUN,${EASYJSON_GITHUB}) easyjson -build_tags opencontrail $<
+
+.easyjson: flow/flow.pb_easyjson.go $(GEN_EASYJSON_FILES_TAG) $(GEN_EASYJSON_FILES_TAG_LINUX) $(GEN_EASYJSON_FILES_TAG_OPENCONTRAIL)
```
Comment (hunchback) [inline Makefile]: But in the case when `WITH_OPENCONTRAIL=false` I don't think you should build `routing_table_easyjson.go`

File: Makefile [hunk 4]
```diff
@@ -211,21 +205,36 @@ debug.analyzer:
 	sed -e 's/type Flow struct {/type Flow struct { XXX_state flowState `json:"-"`/' -i.bak flow/flow.pb.go
```
[the remark this answers is not in the document]
Comment (lebauce) [reply Makefile]: Can you please put the full line ? :-)

**Comment by hunchback** [comment]
sed -e 's/type Flow struct {/type Flow struct { XXX_state flowState
`json:"-"`/' -i.bak $<

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (lebauce) [reply Makefile]: done

File: Makefile [hunk 5]
```diff
@@ -244,12 +253,18 @@ BINDATA_DIRS := \
 npm.install:
 	cd js && npm install
 
-.PHONY: typescript
-typescript: npm.install
+.PHONY: .typescript
+.typescript: npm.install
 	cd js && PATH=`npm bin`:$$PATH make all
 
+.PHONY: .typescript
```
[the remark this answers is not in the document]
Comment (lebauce) [reply Makefile]: done

File: Makefile [hunk 6]
```diff
@@ -244,12 +253,18 @@ BINDATA_DIRS := \
...
+.typescript.clean:
```
[the remark this answers is not in the document]
Comment (lebauce) [reply Makefile]: done

File: flow/flow.proto [hunk 7]
```diff
@@ -176,3 +176,13 @@ message Flow {
...
+
+message FlowSet {
+  repeated Flow Flows = 1;
+  int64 Start = 2;
+  int64 End = 3;
+}
+
+message FlowSearchReply {
+  FlowSet FlowSet = 1;
+}
```
[the remark this answers is not in the document]
Comment (lebauce) [reply flow/flow.proto]: I'm sorry, I don't see it

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (hunchback) [reply Makefile]:
```
sed -e 's/type Flow struct {/type Flow struct { XXX_state flowState `json:"-"`/' -i.bak $<
```

File: flow/flow.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (hunchback) [reply flow/flow.proto]: my mistake

File: Makefile [hunk 8]
```diff
@@ -211,21 +205,36 @@ debug.analyzer:
...
 	gofmt -s -w flow/flow.pb.go
 
-.PHONY: .easyjson.all
-.easyjson.all: builddep .proto ${EASYJSON_FILES_ALL}
-	$(call VENDOR_RUN,${EASYJSON_GITHUB}) easyjson -all ${EASYJSON_FILES_ALL}
+.proto: govendor flow/flow.pb.go filters/filters.pb.go http/wsstructmessage.pb.go
+
+.PHONY: .proto.clean
+.proto.clean:
+	find . \( -name *.pb.go ! -path './vendor/*' \) -exec rm {} \;
```
Comment (nplanel) [inline Makefile]: rm -f flow/flow.pb.go filters/filters.pb.go http/wsstructmessage.pb.go

**Comment by nplanel** [comment]
run skydive-packaging-tests

File: Makefile [hunk 9]
```diff
@@ -40,9 +37,6 @@ SKYDIVE_PATH:=$(SKYDIVE_PKG)/src/$(SKYDIVE_GITHUB)/
 SKYDIVE_GITHUB_VERSION:=$(SKYDIVE_GITHUB)/version.Version=${VERSION}
 GO_BINDATA_GITHUB:=github.com/jteeuwen/go-bindata/go-bindata
 PROTOC_GEN_GO_GITHUB:=github.com/golang/protobuf/protoc-gen-go
-FLOW_PROTO_FILES=flow/flow.proto flow/set.proto flow/request.proto
-FILTERS_PROTO_FILES=filters/filters.proto
```
Comment (nplanel) [inline Makefile]: don't remove these, they are used in SKYDIVE_PROTO_FILES
you should add _easyjson.go in SKYDIVE_PROTO_FILES too
currently "make dist localdist" failed

File: Makefile (hunk 3 above)
Comment (lebauce) [reply Makefile]: That brings other problems: when we generate the tarball (for the release, the packaging, ...), we also need to put WITH_OPENCONTRAIL=true. And I'm sure we'll forget to do so.

So I prefer to stick with this approach

File: Makefile (hunk 9 above)
Comment (lebauce) [reply Makefile]: done

**Comment by lebauce** [comment]
run packaging-tests

**Comment by nplanel** [comment]
run tests

**Comment by lebauce** [comment]
run skydive-k8s-tests

File: Makefile (hunk 3 above)
Comment (hunchback) [reply Makefile]: fine

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by nplanel** [comment]
run packaging-tests

**Comment by lebauce** [comment]
run run skydive-functional-tests-backend-orientdb

**Comment by nplanel** [comment]
run skydive-packaging-tests

**Pull request merged**